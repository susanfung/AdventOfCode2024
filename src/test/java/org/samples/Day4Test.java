package org.samples;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import java.io.IOException;

public class Day4Test {
    @Test
    public void testDay4Part1_findXmasWithMarkers() throws IOException {
        Day4 day4 = new Day4();
        String filePath = "src/test/java/org/samples/Day4TestData.txt";

        String result = day4.findXmasWithMarkers(filePath);

        Approvals.verify(result);
    }

    @Test
    public void testDay4Part1_answer() throws IOException {
        Day4 day4 = new Day4();
        String filePath = "src/test/java/org/samples/Day4Data.txt";

        String result = day4.findXmasWithMarkers(filePath);

        Approvals.verify(result);
    }

    @Test
    public void testDay4Part2_findXMas() throws IOException {
        Day4 day4 = new Day4();
        String filePath = "src/test/java/org/samples/Day4TestData.txt";

        String result = day4.findXMas(filePath);

        Approvals.verify(result);
    }
}
