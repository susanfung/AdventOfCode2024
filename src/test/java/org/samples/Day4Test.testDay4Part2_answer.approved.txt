Found 1877 X-MAS patterns:

......MMMSMMS.M...........................M.M.............M.S.M.....M.M......S.SS.M................................S.M.S.M...............M.M
SMSS.M.AA.AA.A.............................A.....S.M....S.SASAS.M.M.SAM.......A..A........S.M.S.S.S.S.S.............A.A.A.................A.
.AA.A.SMSSMMS.M..M.S.M....................S.S.....A......AMAS.MA...ASAS......M.MS.M........A...A.A.A.A..........M.MS.MMSSM..S.S.....S.M.SSMS
MMMS.MMAM.........A.A........................MMSMS.M....M.MAMAS.M.M.S.M.M........A........S.MSMSM.M.MSMS...M.M...A.....A.....A....S.MA.AMAM.
......MAMS.S.....M.S.M.M.....SMSM..S.S........AA........SAM.SAMA...A...A........S.M....M.M....AAMAMAMAA.....A...S.S...M.S..MMSM....AS.M.SAM.
..M.M.S.S.A.M.S.M.....A.......AA....A........MSSSM......SASAMASMMSMMS.S.S.....M.M.......A....MSMSASASMSM...S.S...A..........A.....S.MM.MSMSS
...A.....M.M.A.A...S.S.SS.S..MSMS..MSMM.......A.A.......SAM.SAM.A.A....A.......A.......S.S.....AS.S.SA......A...M.M........M.S........A..AA.
..S.S.......M.S.M...A....A...........A.....S.M.S.M.S.M..MAM.M.SM.S.M.MMSMM.S.MSSSMSS....AS.S..M.MA.AM.M....M.M.M.M..M.M.S.M......S.M.S.SMMMS
...A.........A.M.S.M.M..M.M.S.M.....S.M.....A.......A...S.M.M.M.A.A...A.A.A.A.AAAAA....M.MA....AM.M.M.M.......ASAS...A...A........ASAM..A.A.
..M.M.S.....M.S.A..........A.A.............S.M.....S.M....MASA.M.S.M.M.S.M.S.MMSMMMS.....MSMMMSMS....A..S.M..S.SAS..S.S.S.M......S.MAS.S.M.S
SSMM.A...S.S...M.S........M.S.M...M.M.........S.S..M.M....SAS.S.AM.M.MA.........AM.M.......AA.AA..S.S.S.SAS...AM.M.......A.........S.M......
.AA.M.S...A...................S.S..A...........A....A.....M.S..M.SA.AM.S.S.S.S.S.MA......MSSMSMSM..A.A.ASAM..M.M......S.S.M..MSSMM..A.......
SSMMSA...M.M.............M.M...A..S.S.........M.MSMSSS.S...A....AS.S.S..A.A.A.A..S.S.M.S..A.......M.M.M.M.M..M.S.MS.M..A......AAA.SSSMS.....
...AM.S..........M.S.M.S.MA...M.M.MAMM.M..........AAA.A...M.S.SMMS.MAS.MSMSM.M.M......A..M.S.S.M.M.......A....A.A..A..M.MS.S.MSSMM.A.A...M.S
S.M.S.........S.M.A.A.A.AS.S...A.AMAM.A.....S.M.MMMMSMSM..M.S..AA.AM.MA.MAM....MSMSSSM.S....A...AS.S..S.S.S..M.SMMS.M.....A.......M.M.M...A.
.A.............A.M.S.M.S.M..M.S.S.SMSS.S.....A...A...A.....A..SMMS.M.S.MMAM..S.MAA.AA.A....M.M.S.SA....A.A...M.S.A.A.....M.MS.M..M.M.....M.S
SSMS......M.M.S.M....M.M.M...A......A.......S.M.S.S.MSMS..M.S......MASA.S.S...ASMSMSMM.S.M..A...AM.M..M.M.M...A.M.S.M........A....A.........
..A........A..........A.A..SMMSSMM.M.S...M.M.......A..A....A.......MAS.M...M.S.M........A..S.S.M.M.....A.....M.S.A..........S.M..S.S........
.M.M.M....S.S..M.SSMSS.S.S..AAAAA...A.....A...S.MSSSMM.M..M.S......M.S.M....ASAS.....M.S.M..A.M.M.....S.S.......M.S.....M.S.M.M...AS.M......
.MASA........M.SA.AA.......SMMSSMM.M.S...SMSM..A..A.................A.A....M.SAM......A.A..M.M.A.......A..........M.M....A...A...M.MAM.S....
.SAS.S........AM.SMMM.....................AA..S.MM.M...S.S......S.SM.S.M.....MAM.....M.S.MS.M.S.S.....MMMS.M.......A....M.S.S.S.S.MS.MA.....
.MASA........M.S.M...M.M.....S.S.S.......MSMS...........A........A.........S.S.M......A....A...A........A.A....M.MS.S.M....A.A...A..AM.S....
.M.M.M........AMAM...MAS......A.AM.M.............S.S...M.M......M.MM.S......A........M.SMSSSM.M.M.M.S..M.S.M.S.MA....A....M.M.M.S.MS.M......
.SAS.........M.SAM...SAS.....M.M.MA..........M.M..A.M.S.M...........A......M.M...........AA..S.S.S.A.......S.MAS.S.MSSM.............A.......
.SAS..........AS.S...MAS...S.S...S.S..........A..M.M.A.A........S.MM.S...M.SSMS.........MMSM.MA.A.M.S.......ASAM....A.........MSSM.S.M......
MMMM...M.M.M.M.S.....M.M.S.SAM.S.............S.S....M.S.M.......MAMS.M....A.AA..............AM.M.M........MSMMMS...M.S...S.M...AA......M.S..
MAM.....A...AS.M.S......A..MAMA..SSSS......S.M.......A..S.SM.M..SAM.A.M.SM.SMMM.M.M........S.S.............A.A....S.M....MAS..MSSM......A...
SAS....S.S.S.SA.A..M.S.M.S.S.M.S..AA..M.S.M.A..S.S.SM.S..A..A...S.SS.M.A.........A....M.S...A......S.S...MSSS.S....A.....SAM...........M.S..
S.S.....A....S.M.S..A.A.A........MMMM..A.A.S.M..A.A.....M.MS.S........M.S.M.S.S.S.S....A...M.M.M.M.MAM....A......SSMM....M.SMMS.S.S.S...A...
MAS....M.M.........M.S.M.S.M.S..M.SA..M.S.M.A..MSMMM...S.S.......S.S...A.A...A...A...MMMSM.S..A.ASAMAM.S.M.S......A...M.M...AAM.MAMA...M.S..
MAM...S.S....M.M.M.S........A...SAS.S.SAMAMS.M..AA......A.........A...M.S.M.M.M.M.M...A...AM.S.S.SAS.SA..........S.MM.SA...SMMSAMAM.M..M.M..
M.S....A......A...A.....M.MM.S..MASASAM.SAM.M..SSSM....M.M..M.S..M.M.........A.......S.S.M.SA...ASAM.M.S.S.M.........AS.S...M.S.S.S.S...AM.S
......M.M....S.S.M.S..S.SA..AM.MM.M.MAMAS.SA.........M.S.....A..........M.M.S.S.S..........M.SSMMM.M......A.........M.S.S.S..A.....A...S.SAM
.S.M...A......A........ASMSM.SAM....S.MAS.S.S.........A.....M.S.......M.SAM....A...............A.........S.M..S.S.S.M.SA.A..M.S...M.M....MAS
..AS.MS.S....M.M.S.M..M.MMAM.SAS.S.M..MAS........S.S.M.S...............ASAS.S.M.M..S.S.S.M....S.M..............A.A...AMMMSMM...M.S.......S.M
.S.MA..A.....M.S.MA....S.SAS.S.M..A...M.S.S.S.S...A..............SMMS.M.SAMA........A...A.M.S.SAM........S.S..M.M.M.M.S.A.A.....A........M.M
.M.S.MM.M.....A.AS.M....ASAS.....S.M.......A.A...M.M..............AA.AS.M.M.M....S.M.M.S.M.A..SAM.M.......A............M.S.M...M.S..S.S...A.
..A.M.S......M.S.M.....M.M.M......A.......M.M.M...............MSSSMMS.MA..........A......MMSSMS.MAS.S..M.M.M.S.S...M.S.M..A..........A...S.S
.M.S.A.......SAS.S.......SMSMS...S.M.....M.S.M......S.M........AA..AMAS.M.....S.SS.M...M.MA.AA..S.SA....A.A...A.....A.AM.S.M.MM.MS.SMSM...A.
....M.S.MM.M.MASA.........AAA.............A.A........A.....S.SMMSMM.SAMA.S.S...A........AM.SMMM..AM.M..S.S.S.M.MM.MM.S.MA...A..A..A.A...MMSM
....SASA..A..M.M.M.......MSMSM.S.M.M.S.M.M.S.M.....SSSMS.M.MA....A..M.S.M.A...MMMM...S.S.SA.....M.MA.......M.S.M.A.....M.SMSMSSSSM.M.M...A..
....MASMMS.SSSS.S.........A.....A...A...A....M.S.M..A.A...AMSMS.S.M.....SMSM....A.....AS.M.S..S.M.S.S.S....SASA.S.S........AA.A........MMSS.
M.S.MAM.A...AA.A.....S.S.S.S.S.S.M.M.S.S.S.S..ASAM.M.M.M.S.SAA.M.M.......A...S.S.S...M.MA......A.....A.....MAS.M..M.M.M.S.SMSM.M.......SAM..
.A..MASM.S.MMMM.MMSMMMASA...A.A.........ASAM.M.SAM.......M.SMSM.A.......M.M...A.AS.S..AS.M....S.M...M.M....MAMA.M.SAMA...A.............MAS.S
M.S.S.S...M.M....AAA.MAM.M.M.MMMS..S.S.M.MAM..AS.M........A....S.S.M.S.......M.M.MA..S.S.M..M.M............M.S.M.ASAS.S.M.M............S.MA.
...........AM.S.MSSSMMAS.......A....A.MSSSSMSM.S....S.S.SM.S........A............M.M....A....A..................M.S.M....A....S.S.SM.S...M.M
.SSMS..M.SS.SA.A.....SAS.S....M.S..M.M.AA..A........MASA..A........M.S.....S.S.....S.S.S.M..S.S.S.....M.M.M........M.S..S.S.M..A.A..AS.S....
..AA.M.SA...MMSSMS.M.MASA......A.....SMMSMM.M.......MAM.MM.SSMSMS...........A.......A.A...S.MA.A.......A.AS.S..S.M..A......A..M.M.MM.SA.....
SSMMM.AM.S....A...A..M.M.M....M.S.....A..........S.MM.SMSS...AAA...........M.M.M...M.M.M..SAM.M.M.....S.S.SA....A..M.S....S.M.M......M.M....
.AA..M.S.M.M.M.S.S.M.S.S.SM.S.M..MMSMS.M..........A.M.MAA...MSMSM......S.M.S..A.......S.M.SAMM.S...S.M.A..M.M..S.M...........A.......M.M....
SSMM.....MAS.MAS......A.A..A.A....AA.............S.M.AMMMS...AA.........A.AM.S.S.......A..S.M.A.....A.M.M...................S.S.M.....A..M.M
.AA.S.S..SASAMAS.M.S.M.M.MM.S.M.SMSSS.M.M...........S.S.....SMSM.......S.M.SA.........S.M.S.MMMS...S.M.S.S.......S.M....M.S.M.SAS....S.S..A.
SSMM.A...M.SAM.S..A..............A.....A....SMSMS.M.MA.....................M.S...........A...A......A...A.........AS.S...A.A..SAM.......MSMS
.A..MMMS..AS.MA..M.S.....M.M.S.MM.MS.SS.S.M..AAA...AM.M...............S.S...AM.S..M.MMMMM.S.S.S.S.SSSM.MSMM.S.M.SS.MA...M.S.M.M.M........AA.
S.M.M.A..M.S.M.S.M...S.S.SA..SAM....A....A..MSMSM.S.S........M.M.......A...M.SA....A.AAA...A...AMAMA....AA.A.A.A...M.M.....M.M.M.M..S.M.SMSM
...A.M.S...M.SA.A....SAMAS.S.SAM...M.M..S.M....A...A.........SAS...S.MMMMSM..M.S.SS.SSSSS.M.M.M.MAM.M..SSSM.S.M.S...A.......ASASAS...A......
.SSMS.AS.S.MAM.S.M...MAM.MA..S.M..............S.S.M.M..M.S.M.SAS....A...AA......A..A.A....S.M...S.S..........A.S.M.S.S.....S.SASAS.SSSM.....
..A..M.SA..MAS..A..S.S.M.MMMM.M..M.M....................A...AM.M..SSSM.MSSS..S.M.MM.M.M....A................S.M.A.AM.M.MS.MSMMMMAMA.A.......
.S.M...M.M.S.S.S.M..AMMMM..A.AS.MSAS..S.M...........M.MMMSMM.S.....A.....A...MAS..........S.M.....S.SM.S.......S.M.SA.A..A..AA.MAM.M.M......
....S.S....M.M.MA..M.MAA..S.S.SAMSAS...A......M.M....A.A.A....M.S.M.M.S.M.M..SAM............MMSSM..A..A..M.M.......S.SMSSSMSSMSSSSSMSMS.....
.....A......ASAS.MM.SSSSS..A..SAMM.M..S.M......AS.M.S.S.S.S...SAM...MASA.....M.S.S....S.M....AAA..M.MM.S.SAM...........AAA..AA.A.A.AAA......
....M.M....S.SAS..SAM.AA..M.M.S.M......A...S.MS.SA.....A......MAS...MAS.M.......A.....MAS.M.MMSSM..S.S...SAS......MMSSMMSMMSMMM.M.MSMSM.....
....S.M.....AM.M..MASMMMM...S.M......MSSMM..A...S.M...M.M.....S.M.M.M.S.S.M.S..M.M....SAMAM.........A....SAMM.M.M.MAAA.A.......M.S..AA......
M.S.MA.....M.M...MSMM........A...M.M.MA.A..S.M...................A...A..MASA....MSSS..M.SAM........M.M.S.M.M.A.ASAMMSSM.S..M.M.SAS.MSMS.....
.A.AS.M...........A.........S.M...A.AM.S.M...........M.S........S.S.M.S.SAM.S.S..AA.....SASM.M.S.S....AS.S.SS.S.SAS..A......A..MAS..........
M.S.MA..M.M......S.SM.S..........S.S.S..A....M.S.M....A......S.M...A....M.S..A.MMMSMM.S.S.M.A...A..M.M.SA.A....AS.M.S.M.S..S.S.M.M..........
...AS.M..A...........A...MSMS......S.S.S.M....A.A....M.S......A..MSSM.....S.MSMMA....A.....S.S.M.M..A..M.M.M.SMMM....A.A........AS.S........
..S.M...S.S....M.S..M.S...AA.....S.SASAM.....M.SMMS...A......S.M..A.......SAM.AS.S..M.S............S.S......A.A.....S.M.SS.S...S.SA....S.S..
..........M.MS.SA........SMSM.SMMSAM.MAM.S....A.AA...M.S.........M.S.S.M..SAMS.MA........................S.M.S.M.........MAM.M.M.M.M....A...
.S.S......MAM.AM.S........A..A.AAM.M.S.MA....M.SMMS.......M.S.M.......A...S.M..MSMM..........MSSS.M.S.M...A..............MAMA.A.A....M.M.M..
..A.......SASM.M...SSSSS.MSMM.SMMS..AMAM.S.M.M.S......M.S.MA.A.......S.M.........A............AA...A.A...S.M..S.M.....M.SS.S.S.S.S.S.MASAM..
.M.M......SAS.......AAA....A.A.A...M.SAM..A...A........A.AM.S.M.......A.......S.S.M..........MMSM.M.S.M........A..S.S..A....M.M....MASASAS..
..........M.M.M.M..MMMMMS.S.M.S.M.M..S.S.S.M.M.S.S.M..M.S.M.SA.......S.M.......A.MSMM.S.M.S....A.........M.S..S.M..A..M.S....AM.M..SAM.SAM..
.S.MS.S.M.S.M..A.......AM.S..A..SAS...........A...A.S.MASA.AS.M...M.M.........M.M.AA.A.A.A....S.S.....S.M.A....M.MM.M.S.S...S.SAS..M.SAMAM..
..A..A...A.A..S.S.....M.SA..M.S.SAS..........M.S.S.M.AMAS.M.S.M.S.MA.M.S.M..SMSS.SSSM.S.M.SM.M.A..S.M..A.M.S....A......A.....ASAS....M.S.M..
SSSMMSMSMSS.M..A.M.M.M..M.S.M..AMAM........S.S....A.S.MASA.M.M.A.AS.S.A.AM.M.AA.............ASMSM..A..S.M......S.S....M.MS.SM.M.M.M.S.M.....
.A..AA..A.....MSMSA.A......A..M.M.M.M.M.....A..S.S.M..M.S.M.A.M.S.MA.M.S.MA.MMMS.M.M.S.S.S.S.SA..MSSM.....................A........A.A..SMMS
M.MMSSSMSMM.S...AS.S.S....S.M......A.A.....M.M..A..M.S.....S.S..M.MMMSA..SSSM..M.SA..SAM..AM.M.M..A....................M.M.MM.S.M.M.S.M..AA.
.M.MA...AA.A...M.M..A......A......S.S.S.S.S.S.SM.M..A..M.M..A....A..AM.S..AA....AS.S.MAM.M.MA....M.S.M....S.M........S.SA...MASA....M.M.SMMS
..AM.S.SSSM.S..S.M.M.M..S.S.M......A.A.A.A...A.....M.S.SAM.MSMS.S.SMMSM..MSMM..M.S...S.MM.MS.SSMMSAMAS.....AM.M.M.....AS.S..MAS.M....A...A..
.S.SA...........AS.S.....A........M.MMMMM.M.M.M.S......SAS.M.A.A.....AM.S.M....S.M...SAM.A..A..AAMASAM....S.MA.A.....M.M....M.S....SSMSMSMMM
..AM.S.S.M.S.S.S.MA.....M.M....M.M....A...S.M..A.......S.MA.M.M.M...S.SA.A......A....SAMS.SM.MSMMSAMAS......S.S.S...................A..AA.A.
MMMM....AS.SA....M.M.....A......A....S.S...A..M.S......M.S.M..S.S.....M.S.MM.M.S.M...S.M......A..M.S.M...........................M.S.MMSMS.S
.AA....S.MAM.M..........S.S....SSSS.......S.M.......M.M.A.A....A.......A.A..A................S.SMSMS..........S.S.................A.A...A.A.
SSSSS....MAMAM.S.M.......A....M.MA...................A.M.S.M..M.M.....MMSMMS.S....M.S.S.S........AA......M.S...AM.SMMS.MS.M......M.S.M.M.M.M
...A.....S.S.SA.AM.S.M..M.M.M..AM.M.................S.S.M....MMSM.M....AAA..AM.S...A...A..M.M.S.SMSMM.....A...M.MA.AA.A..A..................
..MSMM.M.M...M.S.MA.A.S.MASAM.S.S....S.S...S.S......S.MAS.....AA.A..MSMSSSMM.MA...M.S.M.M..A...AS.MA.....M.SS.S.M.SMMSMMS.M......S.M........
....A..SAS....A.AM.S.M.ASASAS.S.......A.....A.....M.SASAM..M.MSSS.S..AA.A....M.S......M.S.S.S.M.MAS.M.S......A.........A..........A..S.SM.S.
.SMSSM.SAS...M.S.MAM..S.M.SAMA..S.S.SM.M...M.M.....AS.M.S...A..A.A..SMSMSMS...A...M.S..A..MAS...SAM..A......M.M.......M.S........S.M..A..A..
..AA...M.MMMMSAM.MAS.M.M.MM.M.M.SASA..............M.SAS.S..S.SM.M.M..A.A.A...M.S...A..M.S.MAM.S.M.S.M.S.M...............S.S.S........M.MM.S.
.MMMSMMMS.SAAMAS.S.S..A.A.......MAM.M.........M.S...SAMA............M.M.M.M.......MMSSS...M.SA.A.A...A.A.................A.A...S.S.S.....A..
.....AA..ASMSS.M.S.S.S.S.S......M.M.M.M........A..M.M.M.MS.SSMMS.......M.S.......S.MAA.M.M.AM.SSMSSSM.S.M...S.M.........M.M.M...A.A.....MMSS
M.S.SSMSM.M.M...A...A....M.M.........A..S.S...M.S.MASA.A..A..AA..SMSM...A........SAMMSM.A.M.S...A.AA.........A.M.M....S.M......M.M.M......A.
.A.....A.A.A...M.S.M.M....AS.SMSSM..S.S..A..S.S...SAS.S.SMSMSMMS.MAA..SMMSS......SAM...S.S.A...M.MSMM.......S.M.A......A.....M.SAM......MMMS
M.S...SMSSS.S............S.SA..AA.......M.M.MAS.M.MASA.A.A......AMSMS..A.A.......S.M.S.M..M.S..................S.S...MSSM.....ASAS.......A..
........A...........M.M....M.MMSSM.S.S.S...AM.MAM.MAM.M.M.M.SSSS.M....S.M.S.M.S...A...A...............................A......M.SAM.....MSMS.
.......M.S...........A..............A.A.M.S.S.SAMAM.S.MM.S.M.AA..............A...S.MMSSM.S.M.................SSSSS.S.M.S.MS.M.AMAM......AA..
SSSM....M.M.S.M.....SMSM...........M.M.M.A....S.S.SA.A..A.A.MMMM............M.S.M....A....A...................AAA.A..S.SA..A.M.S.M.....SMSM.
.AA......A...A.......AA........S.S......M.S.S.MA..M.S.MM.S.M.A.M.S.M.........A.A....M.S.SSMM.M..S.M..........MMMMM.M.SAS.MSSMS.M.........M.S
MSMM....S.S.SMMS.M..MSMS........A........A...AM.M..A......A.S.S.A.AS.S.SS.M.M.S.M........A..A....A..................AM.MA.A...A...........A.
..............A.A......M.S.....M.M....S.M.S.S.M...M.S.SM.S.M...M.S.MA.A..A.........S.S..S.MSSSM.S.M..S.S.SM.M......M.SAS.M.S.S.M.....M.S.M.S
.........MSMMMSSMM......A..............A...A.........A..A.........AM.MMMSSM.M.M.S...A......AAA.A......A.A..A.M.S.....SAS.S............A.A...
..........AA.A.A..M.M..M.S...M.S.M.S..S.MMMSMM......M.MMMSMS.S.M.S.M...AA..A...A...M.M....MMSMM.S.M.SMMM.MSSSMA..M.S.M.MA............M.S.M..
.......S.SSSM.S.M..A..........A.A.A.......A.A.....S.S..AAA..AS.SA.A..MMMSSSSSSM.S.M.M.........MASA.A.A..S.MAAM.S..A.A..M.M..................
M.M.M...A.A.M.M.M.S.S...M.M..M.S.M.S.....M.S.MS.S..A..SMSSSM.MAM.S.M..A...A.AA.....A..........MAS.M.S.M.MAMSMM...M.S.M........S.S..M.M......
.A.A...M.M.M.A.A......M.MA......A.A............A..M.M..MAMA..M.MASA..S.S.M.MSMM...S.S.........M.S.....MASAMA.M.S....A..........A...SAM..MSMS
S.S.S.....S.S.S.SMSSM..AS.S.MSSSSMMS.S.S......M.M..M.S.MAS.MS.SMAS.M..AS.S........S.M.................SASAS.SMAS.M.S.M.S...M.MMMM..SASMMMAA.
.A.A..M.M..A.A....AA..S.S....AA..A....A......M.M....A..SAS...A.M.S.S.M.MA..........AMSSM.S.M..........M.S.MA.MASA.....A.....A.A....S.MAASMSM
M.M.M..A..M.M.MM.MSSM...S.M.MMSMS.M..M.M......AS.M.M.S.M.S..M.M...A....M.MS.M.S.S.S.MAA.A.A.........M.S...M.MM.S.M.M.M.S.M.S.S.S.....SSMSA..
......S.S.......A...M.S.MAS......AMMMS....S.MS.SA...AS.SA........M.M.......A...A.AM.MSSM.S.M.........A.....A..A...A.A...A...A.A........AMMMS
.......A.......S.S...A.ASAM.S.M.S.MAA......A.S.S.M.M.SAMMSM.S.MSSMMM.S.MM.S.M.M.M.MA................M.SMSSS.SMSSSSMS.S.S.M.M.M.M...M.MM.MAA.
.S.S.SM.M.M.M...A...M.S.M.SAMA.A..SMSSMSMMS.M.A.....SMMM.A...A.AAA..A.A..A...ASAS.S.S..S.S..S.S.M.S.....AA...A...A....A...M.MS.S.S..A.S.SMSS
..A.A......A...M.M.......AMAS.M.S...A..AA....M.MS.S.SA..S.S.S.MSSMMM.S.MM.S.S.SASAM.M...A....A...AM.M..MMSM.M.M.S.M..M.M...A..A.A..S.S.A.A..
.M.M.M....S.S......M.M..M.SAM......M.SSSSM.......AMAS.M..A...A..A.S.M.........M.M.MA..MMMMMSMMM.M.SA..M.MA..............M.S.SMSM.M....M.M.M.
....A...............A.....MAS.S.M.S...A.....S.S.M.MAMA..M.M.SSMS.M.A....S.S.S.SMMSS.S.SA.A..AA...AS.S..AS.S..........MMMSA...A.........A....
...S.S.....SMMM...MSMS....S.M..AMAM.MM.M.....A....M.S.M.......A...SMMS.M.AMAMA..AA...AS.S.SSSMS.M.SAS.S.S.M.M.........AAM.S.M.M.......S.S...
............AA..M.MAA.........S.MASA..A....SMSMSMS...........MSMSM..A.A.M.MAM.MMMSSMM.MA...A......M.MA.A...A.........SMSS...................
...........SSMS.SASMSM........M.S.SMSS.S....A.AAA..M.S.M.S.M...AA..M.S.M.AS.S.M.A.AA..M.M.M.M.M.S.M.M.M.M.S.S.S.......A..M.S................
SSSM.S.M....AA.ASASAM.M........AS.M.A......M.MSMSM.SAMA.A.A...MSMS....A.S.S..A.M.SSMM......S.S.A.A...AM.MA...A.......M.M..A.................
.AA...A....SMMM.MAM.MA..S.M...M.SAMM.S.S.S........AMAS.M.S.M....A....S.M..M.S.M.A...........A.MSSMM.S.SAM.SMMMM......MAS.M.SMMS.M.SSSMMSMMM.
MSMM.S.M........M.M.S.S..A......SAM...A.A........M.SAM.........M.M.....M.S.A...M.S.S.S.S.S.M.M.AAA....S.S..AAS.S.S...SAS.SA.AA.A...AA..AAA..
.......MMSSMM........A..S.M.....SAM.SMMMSM.....S.MAMAS.M.............S.MA.M.S.....AMAS..A.....MSSMM.S..AS.SSMSA.A....M.SAM.SMMS.M.MSMMSSSMS.
...M.M..AAAA.M.M..S.M.M..A..M.S.S.M..A.A........AM.SAMAS..............AM.S.......M.MAM.M.M.......ASAM.M.MA...M.MSMM....M.M....M.S.M.A..A.A..
.S.SAS.MMSSMM.A....A....S.M..A......S.M.S.SM.S.S.M.M.SAM..S.M.S......S.M...........M.S..A...M.S.S.MAS...M.M......A......S.S.S..A.A.S.MM.M.M.
..ASAS.......S.S..S.M.......M.S......A...A..A........M.SMMMA.A........A..............S.S.S...A....S.M...........S.M.M.M.MA.A..M.S.M.M.S.SSSS
.M.MMMMSMS.S..............M.S...M.M.S.M.M.MM.S....M.M...AAS.M.S.S.S..S.MS.S...........A.....MSSSM.M.S.M...........S.MA.AM.M.M..A...A...A.AA.
S.M..A.AA.A.....S.MM.M.....A.....A.....A.A..A...S.SAS..SSMS....A.AS.M....A...........M.M......AA...A.A.S.S.S..M.M..AS.S.S.M...MMSMS.S.M.MMMM
.AM.S.SMSM.M.....A..A....SMSS.MSSSS.S.SMSSSM.SS.MASAS.M..A....M.M.MA...MMSM................S.MSMM.MSSSM.A.A....A..S.MA...A......AA.A....MSSM
S.MAM..A.....MSMS.MS.S....A....AA....A.AAA.....AM.M.MA..M.M.......S.M..SAM..................A.......A..M.M.M..S.S..MMSM.S.M....SMSM.M....AA.
..MAS.M.M.M...AA...M.S...M.M..MMSM.MMSMMMSMS.MS.MS.SSSM............A...MAS.................S.M.....MSMM.............A....A..............MSSM
..S.S.S.SA...SMSM.S.A...S.M........MAMAMA...A.....A.A.............S.M..S.M..................A.......AA.........M.S.M.S..S.M.S.M.S.MMSS.M....
...A..SAS.S...A..A.M.S...A.........MASAM.S.S.M.M.M.M.MM.S.M.M.M.M.M........................S.M.S.MMSSSMMS.M.....A.A.A......A...A.A.AA.A.....
MSMSM.MAM.S.SMSMM.S.....S.M.SMSM...SASAS..A...A........A...A.A...A......S.S.S..........M.S.M....A.AA..A..A.....MMSMM.SM.S.M.S.M.S.MMSS.MMSSM
.AA.M.M.M..A.A..........M.M..AA....M.SAM.M.M.S.S......M.S.SMSMS.S.S......ASAS...........A.A....S.MMSSSMMS.M......A.....A.A...A...........AA.
SMSM.A....M.M.M..........A..MSMS.M...MAM..A................AAA.......S.SM.MAM..S.MMSS..M.S.MM.M...A..AAA.A......S.S...M.S.M.S.M.........MSSM
....S.S.MS.S......M.S...S.SM.M..A....S.MMSMSM.M......M.M.MMSMSM.......A..AM.M...A.AA......M.SAM..MSSSSMMS.M.M.M..A.....A.....M.M.M.M.M.M.A..
M.M....A.SAM.S.M...A........A..S.M...SAM.A.A.A........AMAS.MA....M.M.M.MS.SM.MMSSMMSSS.M...ASAS.S.MA......M.SA..M.M...M.S.M...A.A..SASASMMSS
.A....S.MMAMA.A.S.M.S.S.S.MS.S.M.S.M.SAMS.S.S.S.M....S.SASAM.MSSMMA.A.A..M.MA..AA....MAM..M.S.M..AMMMS.MMSMAS.S..MMMM.MAMAM.MS.S.S.SAS.SASAM
S.S......S.M.S.M.A...A...A....A.A.A..S.M.......A......AMAS.M...AAS.S.S.S..AS.SMSSM...SAM...A....S.M.A.A.AAM.S.....AA.AMASAMA..A...AMAM.S.MAS
................S.M.M.M.S.M..S.M.S.M..........S.M....M.M.M....SSMM.......S.S.........S.S..M.S......M.S.MSSS......SSSS.S.S.S.SM.M.M.M.M...S.M
